ergate/job.py

```diff
@@ -34,13 +35,26 @@ def mark_running(self, step: WorkflowStep) -> None:
     def mark_failed(self, exception: Exception) -> None:
         self.status = JobStatus.FAILED
 
+    def mark_step_n_completed(
+        self,
+        n: int,
+        return_value: Any,
+        total_steps: int,
+    ) -> None:
+        self.mark_n_steps_completed(
+            n - self.current_step,
+            return_value,
+            total_steps,
+        )
+
     def mark_n_steps_completed(
         self,
         n: int,
         return_value: Any,
         total_steps: int,
     ) -> None:
-        self.steps_completed = min(self.steps_completed + n, total_steps)
+        self.current_step += n
+        self.steps_completed = min(self.steps_completed + 1, total_steps)
```

Comment:
Is it possible here, when `paths` aren't defined and `total_steps` defaults to `len(workflow)`, to end up with `self.steps_completed == total_steps` prematurely, thus marking the workflow as complete whilst it is still actually incomplete?

I think if so, this would only affect the version supporting negative progression.

Reply:
It wasn't usually possible, no, as this method was only called _after_ a step had completed. With negative progression, however, it appears that could probably be the case.